### ISSUE TITLE
async functions in cito

### ISSUE THREAD

**Issue by trusktr** [joelrich/citojs#23]
With Promise support, I believe this means we'll be able to use `async` functions for `children`, which is really appealing:

```js
async function getText() {
  let data = await something() // a network request perhaps?
  return data.text
}

{tag:'ul', children: [
    {tag: 'h1', children: 'Title'},
    getText()
]}
```

**Comment by joelrich** [comment]
That would be nice indeed. I haven't looked at the async/await specification yet but if an async function really just returns a promise, then this would be very simple to support. In an earlier version I had already promise support built in, but I removed it during a performance refactoring.

Previously I thought it would be best to support all types of asynchronous values (promise, observable, generators...) with the adaptor pattern. This would allow to keep the overhead very low. Unfortunately, it would require to wrap promises with a function call which is not very developer friendly.

A compromise might be to support promises directly as well as an internal async type (for which there are factory functions available for less commonly used async types like observables and generators). It would then automatically convert promises to the internal async type which could have the following signature:

- getCurrentValue()
- subscribe(callback)
- unsubscribe()

For the initial rendering, it would first call getCurrentValue() to check if the value is already available. This allows to skip the .then() microtask of promises if the promise has already been resolved. If not, then it calls subscribe with a (shared if possible) callback. When the callback is called, it renders the dom (maybe in an animation frame).

For updates, it would first check if the original async value itself has changed. If not, it does not have to continue because promises/observables themselves are "immutable". If it has changed, it checks if the value is already resolved. If so, then it compares the value with the previous value (oldAsyncValue.getCurrentValue) and updates the DOM accordingly. If the new value hasn't been available, it has to subscribe to the new async value adaptor (and unsubscribe from the old) and compare the value with the previous value in the callback once it's resolved.

This should be a performant way of supporting promises as well as other asynchronous value types. @barneycarroll, what do you think?

**Comment by trusktr** [comment]
Yep, `async` functions always return a Promise, even if you return a static value. For example, the function

```js
async function foo() {
  return 5
}
```

is essentially the equivalent of this non-async function:

```js
function foo() {
  return Promise.resolve(5)
}
```

I really like the ideas being born in this repo!

**Comment by joelrich** [comment]
Thanks for the clarification. I hate to say it, but I still don't have time for coding. Maybe @barneycarroll is interested in implementing it, but otherwise it has to wait. I am sorry...

**Comment by trusktr** [comment]
It's all good. No time requirements with free stuff, and thanks for what you've already made. :)

**Comment by trusktr** [comment]
@joelrich 

> The promise support is disabled at the moment and will soon be moved to a utility.

Why was the original Promise support removed? Can it be toggled?

**Comment by joelrich** [comment]
@trusktr you would also have to move it. If I remember correctly, then it was related to a refactoring and also because I wasn't particularly happy with how I implemented it.
